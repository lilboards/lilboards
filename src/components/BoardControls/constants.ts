export const MINUTE_IN_SECONDS = 60;
export const SECOND_IN_MILLISECONDS = 1000;
