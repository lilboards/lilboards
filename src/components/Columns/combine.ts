import type { DropResult } from 'react-beautiful-dnd';
import type { Items } from '../../types';

const separator = '\n\n---\n\n';

// https://github.com/atlassian/react-beautiful-dnd/blob/master/docs/guides/combining.md
export default function combine(items: Items, result: DropResult) {
  const sourceItemId = result.draggableId;
  const sourceColumnId = result.source.droppableId;
  const sourceText = items[sourceItemId].text.trim();

  const targetItemId = result.combine!.draggableId;
  const targetText = items[targetItemId].text.trim();

  const combinedText = [targetText, sourceText].join(separator);

  return {
    remove: {
      itemId: sourceItemId,
      columnId: sourceColumnId,
    },
    update: {
      itemId: targetItemId,
      text: combinedText,
    },
  };
}
