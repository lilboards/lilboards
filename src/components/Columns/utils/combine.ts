import type { DropResult } from '@hello-pangea/dnd';
import type { Items, LikesItems } from 'src/types';

const separator = '\n\n---\n\n';

/**
 * @see {@link https://github.com/hello-pangea/dnd/blob/main/docs/guides/combining.md}
 */
export function combine(result: DropResult, items: Items, likes: LikesItems) {
  const sourceItemId = result.draggableId;
  const sourceColumnId = result.source.droppableId;
  const sourceText = items[sourceItemId].text.trim();

  const targetItemId = result.combine!.draggableId;
  const targetText = items[targetItemId].text.trim();

  const combinedText = [targetText, sourceText].join(separator);
  const combinedLikes = {
    ...likes[sourceItemId],
    ...likes[targetItemId],
  };

  return {
    remove: {
      itemId: sourceItemId,
      columnId: sourceColumnId,
    },
    update: {
      itemId: targetItemId,
      likes: combinedLikes,
      text: combinedText,
    },
  };
}
