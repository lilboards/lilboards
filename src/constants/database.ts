export enum DatabaseKey {
  board = 'board',
  boards = 'boards',
  columns = 'columns',
  items = 'items',
  itemIds = 'itemIds',
  likes = 'likes',
  list = 'list',
  lists = 'lists',
  rows = 'rows',
  users = 'users',
}
