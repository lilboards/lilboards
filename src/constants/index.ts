export * from './database';

export const HALF_SECOND = 500;
export const REDIRECT_TO = 'redirectTo';
