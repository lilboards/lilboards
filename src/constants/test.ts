export const BOARD_TEST_ID = 'board_test_id';
export const COLUMN_TEST_ID = 'column_test_id';
export const USER_TEST_ID = 'user_test_id';
