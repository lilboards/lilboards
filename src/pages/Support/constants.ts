export const sponsorLinks = [
  {
    text: 'GitHub Sponsors',
    href: 'https://b.remarkabl.org/github-sponsors',
  },
  {
    text: 'Patreon',
    href: 'https://b.remarkabl.org/patreon',
  },
  {
    text: 'Ko-fi',
    href: 'https://b.remarkabl.org/ko-fi',
  },
  {
    text: 'Liberapay',
    href: 'https://b.remarkabl.org/liberapay',
  },
  {
    text: 'Teespring',
    href: 'https://b.remarkabl.org/teespring',
  },
] as const;
