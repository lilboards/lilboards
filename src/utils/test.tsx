import { render as reactTestingLibraryRender } from '@testing-library/react';
import { DragDropContext } from 'react-beautiful-dnd';
import { Provider } from 'react-redux';

import actions, { resetActions } from '../actions';
import { ITEM_IDS } from '../constants';
import {
  BOARD_TEST_ID as boardId,
  COLUMN_TEST_ID as columnId,
  DATE_NOW as dateNow,
  ITEM_TEST_ID as itemId,
  USER_TEST_EMAIL as userEmail,
  USER_TEST_ID as userId,
} from '../constants/test';
import store from '../store';

import type { ReactNode } from 'react';
import type { Columns, Item } from '../types';

/* istanbul ignore next */
function noop() {}

function StoreWrapper(props: { children?: ReactNode }) {
  return (
    <Provider store={store}>
      <DragDropContext onDragEnd={noop}>{props.children}</DragDropContext>
    </Provider>
  );
}

/**
 * @see {@link https://redux.js.org/recipes/writing-tests#components}
 */
export function renderWithStore(ui: JSX.Element) {
  return reactTestingLibraryRender(ui, { wrapper: StoreWrapper });
}

export function resetStore() {
  resetActions.forEach((resetAction) => store.dispatch(resetAction()));
}

export const updateStore = {
  withBoard() {
    const board = {
      createdAt: dateNow,
      createdBy: userId,
      id: boardId,
      name: 'Board One',
    };
    store.dispatch(actions.loadBoard(board));
    return board;
  },

  withColumn() {
    const column = {
      createdAt: dateNow,
      createdBy: userId,
      name: 'Column One',
      [ITEM_IDS]: [itemId],
    };
    const id = columnId;
    const payload = { [id]: column };
    store.dispatch(actions.loadColumns(payload));
    return { ...column, id };
  },

  withColumns(payload: Columns) {
    store.dispatch(actions.loadColumns(payload));
  },

  withItem() {
    const item: Item = {
      createdAt: dateNow,
      createdBy: userId,
      text: 'Item One',
    };
    const payload = { item, itemId };
    store.dispatch(actions.updateItem(payload));
    return { ...item, id: itemId };
  },

  withLike() {
    const payload = { boardId, itemId, userId };
    store.dispatch(actions.likeItem(payload));
  },

  withUser(email = true) {
    const user = {
      email: email ? userEmail : '',
      id: userId,
    };
    store.dispatch(actions.setUser(user));
    return user;
  },

  withUserEditing() {
    const userEditing = {
      boardId,
      columnId,
      itemId,
    };
    store.dispatch(actions.setUserEditing(userEditing));
    return userEditing;
  },
};

export const getStoreState = () => store.getState();
