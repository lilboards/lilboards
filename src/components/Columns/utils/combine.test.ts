import type { DropResult } from '@hello-pangea/dnd';
import { columnId, dateNow, itemId, userId } from 'test/constants';

import { combine } from './combine';

const itemId1 = `${itemId}1`;
const itemId2 = `${itemId}2`;

const item1 = {
  createdAt: dateNow,
  createdBy: userId,
  text: ' Item 1 text ',
};

const item2 = {
  createdAt: dateNow,
  createdBy: userId,
  text: '\nItem 2 text\n',
};

const items = {
  [itemId1]: item1,
  [itemId2]: item2,
};

const likes = {
  [itemId1]: {
    [userId]: true,
  },
  [itemId2]: {
    [userId]: true,
    [`${userId}2`]: true,
  },
};

it('returns items to update and remove', () => {
  const result = {
    combine: {
      draggableId: itemId2,
      droppableId: columnId,
    },
    draggableId: itemId1,
    source: {
      droppableId: columnId,
      index: 0,
    },
  } as DropResult;

  expect(combine(result, items, likes)).toMatchSnapshot();
});
