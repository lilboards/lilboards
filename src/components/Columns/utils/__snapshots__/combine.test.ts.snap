// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns items to update and remove 1`] = `
{
  "remove": {
    "columnId": "column_test_id",
    "itemId": "item_test_id1",
  },
  "update": {
    "itemId": "item_test_id2",
    "likes": {
      "user_test_id": true,
      "user_test_id2": true,
    },
    "text": "Item 2 text

---

Item 1 text",
  },
}
`;
